CRM/Contact/Import/Form/MapField.php

```diff
@@ -595,4 +594,57 @@ private function isLocationTypeRequired($name): bool {
     return (bool) (Civi::$statics[__CLASS__]['location_fields'][$name] ?? FALSE);
   }
 
+  /**
+   * Assign js to hide hide unused fields.
+   *
+   * The mapper fields use QuickForm hierarchical select and, depending on the
+   * first field selection may offer more (hierarchical) fields. Eg. if you
+   * choose 'phone' for the first field then 2 more fields will give you
+   * the chance to enter the phone_type_id and location_type_id.
+   *
+   * When loading the field we hide these hierarchical fields if they are empty,
+   * by js which is assigned here.
+   *
+   * @param array $defaults
+   */
+  protected function assignJsToHideUnusedFields(array $defaults): void {
+    $js = ["<script type='text/javascript'>"];
+    foreach ($defaults as $row => $mapperValues) {
+      foreach ([1, 2, 3] as $columnNumber) {
+        if (empty($mapperValues[$columnNumber])) {
+          $rowNumber = str_replace('mapper[', '', substr($row, 0, -1));
+          $js[] = "CRM.\$('#mapper_{$rowNumber}_{$columnNumber}').hide();";
+        }
+      }
+    }
+    $js[] = "</script>\n";
+    $this->assign('initHideBoxes', implode("\n", $js));
+  }
+
+  /**
+   * Create the defaults for mapper fields.
+   *
+   * Some fields need a second or third value. These are extracted from
+   * $sel2, $sel3 - the select option lists.
+   *
+   * @param string $key
+   * @param array $sel2
+   * @param array $sel3
+   *
+   * @return array
+   */
+  protected function createMapperDefault($key, $sel2, $sel3): array {
+    $def = [$key];
+    if (!empty($sel2[$key]) && is_array($sel2[$key])) {
+      reset($sel2[$key]);
+      $subkey = key($sel2[$key]);
+      $def[] = $subkey;
```

Comment:
I think "Primary" is set here when the form is first loaded and the mapping is guesstimated.  But after the form is built there is no default set for $sel2 so "Home" is selected as that is the first item in the select list.

Reply:
The options are created via JS here: https://github.com/civicrm/civicrm-packages/blob/96b7c2441b851395176bb275a45e24fbe31e3187/HTML/QuickForm/hierselect.php#L427

Reply:
I've just rebased this - do you see a fix for ^^ @darrick - it was kind of a tangent on this PR  - so I'm also happy to drop that bit out of scope for this PR

Reply:
> I think "Primary" is set here when the form is first loaded and the mapping is guesstimated. 

Yes,  'Primary' is selected because it is first in the php list.  

When you change to a relationship ('spouse of') then the JS `swapoptions()' function is called.  No value is initially set so the dropdown shows the first option.  

The array ordering in php is:
```
"Primary":"Primary",
"5":"Billing",
"1":"Home",
"3":"Main",
"4":"Other",
"2":"Work"
```
However, although the php values are passed in the same order, when the JS function loops through the array to add the select options, it does so based on the ordering of the keys, not the order added, and so the options are ordered as:

```
1: "Home"
​​2: "Work"
​​3: "Main"
​​4: "Other"
​​5: "Billing"
​​Primary: "Primary"
```

To fix the default, we could explicitly pass through a default value and use that when creating the select options in `swapoptions()`

If we wanted to solve the more general issue of making the JS options match the PHP order, I think we would need to assign new key values when we pass from PHP to JS, but then also map back from from JS keys to PHP keys.  That sounds like a lot of effort for something few people are likely to notice!

This looks like it goes well beyond the scope of the initial aim of this PR.